File: lab/pages/scripts/apply-lab-overrides.py
#!/usr/bin/env python3
from pathlib import Path
import shutil

SITE = Path('.pages-site')
LAB_PREPAINT = Path('lab/pages/assets/prepaint.css')
TARGET_PREPAINT = SITE / 'override/core/prepaint.css'
MAIN_CSS = SITE / 'override/main.css'
IMPORT_LINE = '@import "./core/prepaint.css?v=unversioned";\n'
ANCHOR = '@import "./components/category-nav/controls.css?v=unversioned";\n'

if not SITE.is_dir() or not MAIN_CSS.is_file() or not LAB_PREPAINT.is_file():
    raise SystemExit('lab Pages staging context is incomplete')

TARGET_PREPAINT.parent.mkdir(parents=True, exist_ok=True)
shutil.copyfile(LAB_PREPAINT, TARGET_PREPAINT)

manifest = MAIN_CSS.read_text(encoding='utf-8')
if IMPORT_LINE in manifest:
    raise SystemExit('lab prepaint import already present in staged override manifest')
if ANCHOR not in manifest:
    raise SystemExit('category controls import anchor missing from staged override manifest')
manifest = manifest.replace(ANCHOR, ANCHOR + IMPORT_LINE, 1)
MAIN_CSS.write_text(manifest, encoding='utf-8')

print('Applied lab-only first-paint CSS to .pages-site staging copy.')

File: lab/pages/steps/apply-lab-overrides.ts
import path from 'node:path';
import { PAGE_ASSETS, ROOT, SITE, assert, isDir, isFile, read, write } from '../lib/core.js';

const LAB_STYLES = ['prepaint.css', 'performance.css'] as const;
const SHARED_COMPONENT_FILES = [
  'override/components/product-card/content.css',
  'override/components/product-card/image-ratio.css',
  'override/components/product-card/layout.css',
  'override/components/product-card/pricing.css',
  'override/components/section-heading/layout.css',
  'override/components/section-heading/section-heading.css',
] as const;

export function applyLabOverrides(): void {
  const mainCss = path.join(SITE, 'override', 'main.css');
  assert(isDir(SITE) && isFile(mainCss), 'lab Pages staging context is incomplete');

  const manifest = read(mainCss);
  assert(
    !LAB_STYLES.some((name) => manifest.includes(`lab-inline:${name}`)),
    'lab first-paint CSS already present in staged production stylesheet',
  );

  const labCss = LAB_STYLES
    .map((name) => `\n/* lab-inline:${name} */\n${read(path.join(PAGE_ASSETS, name)).trim()}\n`)
    .join('');
  write(mainCss, `${manifest.trimEnd()}\n${labCss}`);

  for (const relativePath of SHARED_COMPONENT_FILES) {
    const source = path.join(ROOT, relativePath);
    assert(isFile(source), `frontend source missing: ${relativePath}`);
    assert(!read(source).includes('html.sc-catalog-prepaint'), `lab prepaint alias leaked into shared component source: ${relativePath}`);
  }
}
